src/providers/tenderly-simulation-provider.ts

```diff
@@ -0,0 +1,324 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { JsonRpcProvider } from '@ethersproject/providers';
+import axios from 'axios';
+import { BigNumber } from 'ethers/lib/ethers';
+
+import { SwapRoute } from '../routers';
+import { Erc20__factory } from '../types/other/factories/Erc20__factory';
+import { SwapRouter02__factory } from '../types/other/factories/SwapRouter02__factory';
+import { ChainId, CurrencyAmount, log, SWAP_ROUTER_ADDRESS } from '../util';
+import { APPROVE_TOKEN_FOR_TRANSFER } from '../util/callData';
+import {
+  calculateGasUsed,
+  initSwapRouteFromExisting,
+} from '../util/gas-factory-helpers';
+
+import { IV2PoolProvider } from './v2/pool-provider';
+import { ArbitrumGasData, OptimismGasData } from './v3/gas-data-provider';
+import { IV3PoolProvider } from './v3/pool-provider';
+
+type SimulationResult = {
+  transaction: { hash: string; gas_used: number; error_message: string };
+  simulation: { state_overrides: Record<string, unknown> };
+};
+
+export type TenderlyResponse = {
+  config: {
+    url: string;
+    method: string;
+    data: string;
+  };
+  simulation_results: [SimulationResult, SimulationResult];
+};
+
+const TENDERLY_BATCH_SIMULATE_API = (
+  tenderlyBaseUrl: string,
+  tenderlyUser: string,
+  tenderlyProject: string
+) =>
+  `${tenderlyBaseUrl}/api/v1/account/${tenderlyUser}/project/${tenderlyProject}/simulate-batch`;
+
+// We multiply tenderly gas estimate by this estimate to overestimate gas fee
+const ESTIMATE_MULTIPLIER = 1.25;
+
+/**
+ * Provider for dry running transactions.
+ *
+ * @export
+ * @interface ISimulator
+ */
+export interface ISimulator {
+  /**
+   * Returns a new SwapRoute with updated gas estimates
+   * @returns SwapRoute
+   */
```

Comment:
You should mention that all classes that implement must set the simulationFailed field to true if the simulation fails.

It should part of the contract of the interface